### PR TITLE
[MUGEN] Add VideoGPT Prototype

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #282
* #284
* __->__ #281

Summary:
- Add VideoGPT builder as a prototype with parameters from [VideoGPT: Video Generation using VQ-VAE and Transformers](https://arxiv.org/pdf/2104.10157.pdf) for the BAIR / RoboNet dataset (page 13 Table A1, A2) 
- This model is used by #282 for unit testing a real GPT generation model with video components.

Test Plan:
```
$ python -m pytest --cov=torchmultimodal/models/ test/models/test_video_gpt.py -vv
============================================================ test session starts ============================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/t2v/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 5 items

test/models/test_video_gpt.py::test_encode PASSED                                                                                     [ 20%]
test/models/test_video_gpt.py::test_decode PASSED                                                                                     [ 40%]
test/models/test_video_gpt.py::test_lookup[in-expected_shape0-38.8214] PASSED                                                         [ 60%]
test/models/test_video_gpt.py::test_lookup[out-expected_shape1--23.4659] PASSED                                                       [ 80%]
test/models/test_video_gpt.py::test_forward PASSED                                                                                    [100%]

---------- coverage: platform darwin, python 3.8.13-final-0 ----------
Name                                            Stmts   Miss  Cover
-------------------------------------------------------------------
torchmultimodal/models/video_gpt.py                39      1    97%


=========================== 5 passed in 4.34s =========
```

Differential Revision: [D38932536](https://our.internmc.facebook.com/intern/diff/D38932536)

### CODE REVIEW

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38932536).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38932536).

File: torchmultimodal/models/video_gpt.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    if latent_shape != vqvae_latent_shape:
+        raise ValueError(
+            f"Latent shape required: {latent_shape} does not match that of VQVAE: {vqvae_latent_shape}"
+        )
+
+    # constructs projection layers
+    in_projection = nn.Linear(in_tokenizer.embedding_dim, d_model, bias=False)
+    in_projection.weight.data.normal_(std=0.02)
+    out_projection = nn.Linear(out_tokenizer.embedding_dim, d_model, bias=False)
+    out_projection.weight.data.normal_(std=0.02)
```
Comment (ebsmothers) [inline torchmultimodal/models/video_gpt.py:79]: Similar comment to the other PR: I'm not sure weight initialization should happen in the builder

File: torchmultimodal/models/video_gpt.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        An instance of :class:`~torchmultimodal.models.gpt.MultimodalGPT`.
+    """
+    # constructs in and out tokenizers
+    in_tokenizer = video_vqvae()
+    out_tokenizer = video_vqvae()
+    num_in_tokens = in_tokenizer.num_embeddings  # codebook size
+    num_out_tokens = out_tokenizer.num_embeddings
+
+    # derived parameters
+    vqvae_latent_shape = in_tokenizer.latent_shape(input_shape)
```
[the remark this answers is not in the document]
Comment (langong347) [reply torchmultimodal/models/video_gpt.py:69]: `latent_shape` is computed by `torchmultimodal.models.video_vqvae.VideoEncoder` and exposed as a public API in `VQVAE` here:

https://github.com/facebookresearch/multimodal/blob/2ed86efc6a0781b3a7c8d1c27d700a04a9676864/torchmultimodal/models/vqvae.py#L58-L65

[the remark this answers is not in the document]
Comment (langong347) [reply torchmultimodal/models/video_gpt.py:69]: > Second question:

Answered in #276 : [comment](https://github.com/facebookresearch/multimodal/pull/276#discussion_r954258451)

Right after this line we check the user supplied latent shape is consistent with that derived from VQVAE.

File: torchmultimodal/models/video_gpt.py (hunk 1 above)
Comment (langong347) [reply torchmultimodal/models/video_gpt.py:79]: Updated in #283 . Will also update the callsite here.

File: torchmultimodal/models/video_gpt.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    """
+
+    # this includes both the upsampling conv filter size and encoder filter size
+    # ideally we should code them as two separate parameters in VideoEncoder
+    encoder_kernel_sizes = ((4, 4, 4), (3, 3, 3))
+    # stride (1, 1, 1) is that of the encoder filter conv layer
+    encoder_strides = ((2, 2, 2), (1, 1, 1))
+    encoder_n_layers = len(encoder_strides)
+    decoder_kernel_sizes = ((4, 4, 4),)
+    decoder_strides = ((2, 2, 2),)
```
[the remark this answers is not in the document]
Comment (langong347) [reply torchmultimodal/models/video_gpt.py]: Moved them to API args for better documentation.

File: torchmultimodal/models/video_gpt.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    d_model: int = 576,
+    n_head: int = 4,
+    dropout: float = 0.2,
+    attn_dropout: float = 0.3,
+    num_decoder_layers: int = 16,
+    use_gpt_init: bool = True,
+) -> MultimodalGPT:
+    """VideoGPT model
+
+    For the momemnt this is a prototype implementation which may not fully commit to the specifications
```
Comment (ebsmothers) [inline torchmultimodal/models/video_gpt.py:38]:
```suggestion
    For the moment this is a prototype implementation which may not fully commit to the specifications
```

Also can you clarify what this comment means? Is it just that we have not run an end-to-end loop with it yet, or is there some refactoring you are anticipating?

Comment (langong347) [reply torchmultimodal/models/video_gpt.py:38]: I will be specific in saying that:
- model arch follows the paper
- parameters are for a particular dataset "BAIR/RobotNet" (the paper has a few other sets of parameters for different datasets: [paper](https://arxiv.org/pdf/2104.10157.pdf) page 13 Table A.1, Table A.2)

I don't think we have to say "we have not run an end-to-end loop" as we haven't claimed it as one of our examples (in `examples/` folder). In other words, we could be implementing components/following parameter specs from a SOTA model without onboarding it as an end-to-end example.

What do you think?

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38932536).

**Pull request closed without merging**